core/src/main/scala/kafka/log/ProducerIdMapping.scala

```diff
@@ -314,84 +298,76 @@ class ProducerIdMapping(val config: LogConfig,
   def lastEntry(pid: Long): Option[ProducerIdEntry] = pidMap.get(pid)
 
   /**
-    * Serialize and write the bytes to a file. The file name is a concatenation of:
-    *   - offset
-    *   - a ".snapshot" suffix
-    *
-    *  The snapshot files are located in the logDirectory, inside a 'pid-mapping' sub directory.
-    */
+   * Serialize and write the bytes to a file. The file name is a concatenation of the last offset
+   * included in the snapshot (exclusive) and the ".snapshot" suffix.
+   */
   def maybeTakeSnapshot() {
     // If not a new offset, then it is not worth taking another snapshot
     if (lastMapOffset > lastSnapOffset) {
-      val file = new File(snapDir, formatFileName(lastMapOffset))
-      writeSnapshot(file, pidMap)
+      val snapshotFile = Log.pidSnapshotFilename(logDir, lastMapOffset)
+      debug(s"Writing producer snapshot for partition $topicPartition at offset $lastMapOffset")
+      writeSnapshot(snapshotFile, pidMap)
 
       // Update the last snap offset according to the serialized map
       lastSnapOffset = lastMapOffset
 
-      maybeRemove()
+      maybeRemoveOldestSnapshot()
     }
   }
 
   /**
+   * Get the last offset (exclusive) of the latest snapshot file.
+   */
+  def latestSnapshotOffset: Option[Long] = latestSnapshotFile.map(file => Log.offsetFromFilename(file.getName))
+
+  /**
     * When we remove the head of the log due to retention, we need to
     * clean up the id map. This method takes the new start offset and
     * expires all ids that have a smaller offset.
     *
     * @param startOffset New start offset for the log associated to
     *                    this id map instance
     */
-  def cleanFrom(startOffset: Long) {
-    pidMap.retain((pid, entry) => entry.firstOffset >= startOffset)
-    if (pidMap.isEmpty)
-      lastMapOffset = -1L
+  def expirePids(startOffset: Long) {
+    pidMap.retain((pid, entry) => entry.lastOffset >= startOffset)
+    deleteSnapshotFiles(file => Log.offsetFromFilename(file.getName) <= startOffset)
```

Comment:
We could have deleted all snapshots and therefore need to update lastSnapOffset?